[FSDP][state_dict][3/N] Change how state_dict utils access attributes in _FSDPState

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* #88636
* __->__ #88635
* #88481
* #87900

**What This PR Does**
_state_dict_utils currently accesses the FSDP states through module. To enable composable FSDP state_dict, these accesses need to go through _FSDPState. module is still required for most APIs as state_dict has to access per-module information.

fegin committed:
[FSDP][state_dict][3/N] Change how state_dict utils access attributes in _FSDPState

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][3/N] Change how state_dict utils access attributes in _FSDPState"


**What This PR Does**
_state_dict_utils currently accesses the FSDP states through module. To enable composable FSDP state_dict, these accesses need to go through _FSDPState. module is still required for most APIs as state_dict has to access per-module information.


[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -61,6 +61,24 @@ def _all_handles(state: _FSDPState) -> List:
+    """
+    Given a module and returns the flat handles that map to this module. If the
+    module is FullyShardedDataParallel, the module._handles will be returned.
+    """
```
I am wondering what you think about using the `_is_composable()` util from `_common_utils.py` instead of `if module == state`.

I can see that `if module == state` encodes the intention explicitly, but at the same time, I can see that enforcing the call to `_is_composable(state)` each time the logic between the two code paths diverges can make it easier to find such divergences (i.e. just search for `_is_composable`).

Basically, the implementation of `_is_composable(state)` performs a weaker check of simply seeing if `state` is an instance of `nn.Module`, but it still might be a good idea to use that check instead. Let me know what you think!

fegin replied:
Sure, let's use `is_composable(state)`.

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -61,6 +61,24 @@ def _all_handles(state: _FSDPState) -> List:
 
 
+@no_type_check
+def _module_handles(module: nn.Module, state: _FSDPState) -> List:
```
nit: I am also wondering how you feel about argument ordering. I recall that every function/method I refactored had `state: _FSDPState` as the first argument. I do not think this makes any difference functionally, but maybe we keep that convention?

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -617,13 +672,16 @@ def _pre_load_state_dict_hook(
+@no_type_check
 @torch.no_grad()
 def _post_load_state_dict_hook(module: nn.Module, *args: Any) -> None:
+    # TODO: get the composable state from module
```
Just FYI, I think the syntax is like:
```
fsdp_state = fully_shard.state(module)
```

For example:
https://github.com/pytorch/pytorch/blob/fca6ed02b91e0d685cc9b8b504bac5d356d31876/torch/distributed/_composable/fully_shard.py#L44

fegin replied:
hmm, this can cause a circular import. Can we pull this to a module that does not depend on FSDP?

awgu replied:
Discussed over chat -- let us discuss this further with the composability devs

awgu reviewed (approved):
LGTM!

fegin committed:
Update on "[FSDP][state_dict][3/N] Change how state_dict utils access attributes in _FSDPState"


**What This PR Does**
_state_dict_utils currently accesses the FSDP states through module. To enable composable FSDP state_dict, these accesses need to go through _FSDPState. module is still required for most APIs as state_dict has to access per-module information.


[ghstack-poisoned]

fegin commented:
@pytorchbot merge
